### ISSUE TITLE
Managing large service fabric application code

### ISSUE THREAD

**Issue by AnisTss** [Azure/service-fabric-issues#1569]
I am looking for a guide for developing large applications on service fabric in terms of managing local development per developer.

We work as a team of 4 developers on a service fabric solution. We use visual studio 2019 & service fabric SDK. The project is constructed of a single service fabric application that contains around 15 services and 70 projects. The build time + local deployment on the dev cluster takes around 15 minutes.

**I want to reduce the build + deploy time because it is becoming a frustration for my team developers**

Note 1: The refresh application mode is enabled but it is not consistent. Sometimes VS2019 deletes the app and re-deploys it.

I opened a stackoverflow question [here](https://stackoverflow.com/questions/57578139/developing-large-service-fabric-application) and someone suggested that we should consider splitting the application into multiple small applications as a solution. **Is this a recommended approach ?**

**Comment by masnider** [comment]
TLDR: The tooling makes this a little difficult today, but the overall model has a lot of flexibility. Getting to good unit tests and mocks for calls to other services can do wonders so that you don't have to deploy everything (or anything) when you want to test self-contained changes.

There's a bunch of different options here depending on what you're really trying to do and how intertwined the services are. 

Just splitting stuff up as suggested in SO isn't really a great solution if the services depend on each other. In those cases you want to retain the notion of an application so that you can share config and parameters and an upgrade boundary, etc. It would certainly solve the local deployment speed thing, but also removes some of your ability to code, test, and deploy things as a unit.

The pattern that's the "simplest" is to ensure that each service has mocked out all dependencies on the other services and that small shims send back reasonable responses if you run one service without the other. That way if you're running the whole app locally to test your component, you don't actually need to spin up _all_ the services in order to test your changes. Generally moving to a world where you don't have to spin up the services at all for most unit testing is also preferable. You should only need to spin up the local cluster for failover/chaos testing, and only deploy all the services when you need to do integration testing. 

The existing tools make this a little difficult to achieve (@BigMorty fyi) since they add all the services as default services. But you can strip default services out and instead rely on [Start-Service.ps1](https://github.com/masnider/service-fabric-dotnet-data-aggregation-v2/blob/master/DataAggregation/Scripts/Start-Service.ps1) to ensure that only the services that you care about starting get started when you F5. (Don't worry about this example, it's more complicated than it needs to be :) You just need the create-service calls at the bottom, really.)

This can also lead to divergence where the manifest that gets pulled out of your source control and deployed is different from the one you need for local debugging. You can fix that by removing the dependencies on default services generally, and instead calling create explicitly for new services when necessary, but that can be a bit of a shift for some folks. 

Fundamentally a lot of this comes down to the tooling. As a proof of concept I built a [script](https://github.com/masnider/SFAppPackageGenerator) that builds an application package and app manifest from a bunch of built service packages. This would end up in your CI/CD chain somewhere and would take all the individual packages made by the individuals working on different services/applications into the "real" application package that would get deployed. This lets you split the development into whatever units or chunks you want, while preserving the overall application concept when you actually go to deploy.

Hope this helps a little
-Matt
